core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -132,16 +131,14 @@ public void processExistingRecoveries(RoutingAllocation allocation) {
     public void allocateUnassigned(RoutingAllocation allocation) {
         final RoutingNodes routingNodes = allocation.routingNodes();
         final RoutingNodes.UnassignedShards.UnassignedIterator unassignedIterator = routingNodes.unassigned().iterator();
-        MetaData metaData = allocation.metaData();
         while (unassignedIterator.hasNext()) {
             ShardRouting shard = unassignedIterator.next();
             if (shard.primary()) {
                 continue;
             }
 
             // if we are allocating a replica because of index creation, no need to go and find a copy, there isn't one...
-            IndexMetaData indexMetaData = metaData.getIndexSafe(shard.index());
-            if (shard.allocatedPostIndexCreate(indexMetaData) == false) {
+            if (shard.unassignedInfo().getReason() == UnassignedInfo.Reason.INDEX_CREATED) {
```

Comment:
we're iterating here over unassigned shards, for which we guarantee presence of an unassignedInfo object:

https://github.com/elastic/elasticsearch/pull/19516/files/4e4108cfd053e842d7d0af80b90f20c837f1a7a2#diff-2efe5c966a8d261c9c065dcaf1f93d43R82

I can assert it here as well again if you like.